Add support for transform-box CSS rule

Cu3PO42 opened the pull request:
Hi! This is a PR implementing the functionality I requested in #98.

Supporting this was indeed fairly straight-forward with most changes boiling down to forwarding the correct arguments. The only "real" change is in `HasGeometryContextImpl#transformOrigin`.

I have not yet added any automatic test cases because I'm not sure how to best proceed on that front. As far as I can tell, the majority of tests seem to compare rendering results between JSvg and Batik; however, Batik doesn't support transform-box, so such a test does not seem useful.

I have manually verified that the results fall within my expectations, though.

weisJ commented on jsvg/src/main/java/com/github/weisj/jsvg/nodes/prototype/impl/HasGeometryContextImpl.java:
```diff
@@ -23,52 +23,62 @@
-            @NotNull Length transformOriginY, @Nullable ClipPath clipPath, @Nullable Mask mask,
-            @Nullable Filter filter) {
+            @NotNull Length transformOriginY, @NotNull TransformBox transformBox, @Nullable ClipPath clipPath,
+            @Nullable Mask mask, @Nullable Filter filter, @NotNull Object node) {
```
It should be possible to take an `SVGNode` here instead. I think the fact that `ElementBounds` takes an `Object` parameter is a left over artefact from when it was used for non-related classes, where the nearest superclass was `Object`.

weisJ commented on jsvg/src/main/java/com/github/weisj/jsvg/nodes/prototype/impl/HasGeometryContextImpl.java:
```diff
@@ -92,9 +102,21 @@ private HasGeometryContextImpl(@Nullable AffineTransform transform, @NotNull Len
+        }
+
+        @NotNull ElementBounds bounds = new ElementBounds(node, context);
+        @NotNull Rectangle2D box = transformBox == TransformBox.FillBox ? bounds.boundingBox() : bounds.strokeBox();
```
Let's be explicit here with a switch. Also to future proof (as the `boundingBox()` method is used internally and currently doesn't follow spec semantics) let's introduce a `fillBox()` function on `ElementBounds`. For now it could simply call through to `boundingBox()`.

weisJ reviewed (changes requested):
Looking good. Just a few comments. Thank you :)

weisJ commented:
Running `./gradlew spotlessApply` will fix the CI.

> As far as I can tell, the majority of tests seem to compare rendering results between JSvg and Batik; however, Batik doesn't support transform-box, so such a test does not seem useful.

There are some tests , which compare to a simpler "flattened" reference svg where the attribute isn't used (e.g. in `GeometryAttributesTest.testPathLength()`.

Cu3PO42 commented:
I appreciate your time! I could have sworn I ran the formatter, but I must have made some changes after that. I made the following changes:

* I added a test based on the same concept as `testPathLength`.
* I changed how ElementBounds are passed to HasGeometryContextImpl so that they are duplicated less. `ShapeContainer` and `Use` may still create some duplicates, though.
* I introduced a switch-case to handle the different values of TransformBox
* I added the fillBox method on ElementBounds.

I also realized this implementation of `transform-box` isn't quite complete. The property does not apply to just `transform-box`, but also `transform` itself. With it, it is valid to have something like `transform: translate(-50%, -50%)`, but in order to enable that, the `parseTransform` function must first be updated to handle percentage values.

weisJ commented:
> * I changed how ElementBounds are passed to HasGeometryContextImpl so that they are duplicated less. `ShapeContainer` and `Use` may still create some duplicates, though.

I probably should have more clear in that I don't think duplication of `ElementBounds` needs to be resolved right now. While your solution works it somehow doesn't feel quite right that the caller of `transformOrigin` has to have knowledge about the size of an element. In particular it leads to brittle code, as usually you'd always want/need to pass the `ElementBounds` corresponding to the node you want to call a method on. I'd rather like to have the nodes themselves hold an instance of `ElementBounds` which is used for computation caching. However this will need some bigger refactoring to accommodate (in particular to have cheap cache invalidation). For now I think it is ok to go back to your previous version. Though I appreciate the initiative :)

> I also realized this implementation of `transform-box` isn't quite complete. The property does not apply to just `transform-box`, but also `transform` itself. With it, it is valid to have something like `transform: translate(-50%, -50%)`, but in order to enable that, the `parseTransform` function must first be updated to handle percentage values.

I am currently working on supporting animations for transforms. For this transforms need to be stored as a list of simple transforms. I believe it will make supporting this a lot easier. For now not supporting it should be enough.

Cu3PO42 commented:
No worries, you were sufficiently clear. I just had some time and figured I might tackle that deduplication. I do agree, though, it didn't feel right, but I also didn't want to undertake an even large scale refactoring without really knowing my way around the code base.

I removed the commit the offending commit.

> I am currently working on supporting animations for transforms. For this transforms need to be stored as a list of simple transforms. I believe it will make supporting this a lot easier. For now not supporting it should be enough.

That's fair. Since percentage values aren't supported at all right now for translates, I also agree that not handling them in the context of `transform-box` isn't an inconsistency that is necessarily confusing.

weisJ commented:
Now everything is looking great 🚀 One last thing: Could you please also add a test for the `stroke-box` value?

Cu3PO42 commented:
Sure thing! I added one test. I am not 100% sure the semantics of stroke-box make sense to me, but I verified the reference image's correctness using Firefox and Chrome.

weisJ commented on jsvg/src/test/java/com/github/weisj/jsvg/TransformTest.java:
```diff
@@ -43,4 +43,20 @@ void refTest() {
     }
+
+    @Test
+    void testTransfromBoxFill() {
```
```suggestion
    void testTransformBoxFill() {
```

weisJ commented on jsvg/src/test/java/com/github/weisj/jsvg/TransformTest.java:
```diff
@@ -43,4 +43,20 @@ void refTest() {
+    }
+
+    @Test
+    void testTransfromBoxStroke() {
```
```suggestion
    void testTransformBoxStroke() {
```

weisJ reviewed (changes requested):
Thank you. One last knit. Then I think this will be ready to merge :)

weisJ commented:
Thank you very much for your contribution!

Cu3PO42 commented:
Thanks for you help getting it merged!

Cu3PO42 commented:
Frohes neues :champagne: !

I'm sorry to trouble you, but I wanted to ask if you have any timeline for a next release that might include this feature? No worries if not, I'm just wondering if I should switch to snapshots in a different project that requires this feature.

weisJ commented:
> Frohes neues 🍾 !

Danke sehr. Ebenfalls eine frohes Neues 🎆 

> I'm sorry to trouble you, but I wanted to ask if you have any timeline for a next release that might include this feature? No worries if not, I'm just wondering if I should switch to snapshots in a different project that requires this feature.

I’m trying to release a new version within the next 1-2 weeks. There are some open remaining issues  with the animations support I have to iron out first. Shouldn’t take too long though.
